### DISCUSSION TITLE
Task: Streams and events in RPC

### DISCUSSION THREAD

**Discussion by pfrazee** [atek-cloud/atek#4] [Core Development]
Atek currently handles RPC using JSON-RPC over HTTP. This lacks a mechanism for streams to/from an endpoint, and also lacks a good mechanism for subscribing, unsubscribing, and delivering events. Events in particular are important.

**Comment by pfrazee** [comment]
One possible solution is a separate mechanism for opening streams between services, which would (initially) be a WebSocket. This would require a separate protocol than JSON-RPC, as JSON-RPC lacks semantics for opening a stream.

Events could then be handled using JSON-RPC notifications. This would provide a simple means for registering and unregistering subscribers as the connection's heartbeats will detect disconnects and can unregister the subscriber accordingly.

**Comment by fczuardi** [comment]
[gRPC](https://grpc.io/) has [streams](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc), have you considered it?

**Reply by pfrazee** [reply]
Yeah but the gRPC stack is a bit of a pain to work with. It's really glued to http2 (to the extent that it's a PITA to use our custom networking) and you have to run codegen from the schemas.

Not resolved against it yet, just not in love with it
